flink-runtime/src/main/java/org/apache/flink/streaming/api/operators/StreamOperatorStateHandler.java

```diff
@@ -244,6 +244,11 @@ void snapshotState(
                                 && ((AbstractKeyedStateBackend<?>) keyedStateBackend)
                                         .requiresLegacySynchronousTimerSnapshots(
                                                 checkpointOptions.getCheckpointType());
+                requiresLegacyRawKeyedStateSnapshots |=
```

Comment:
I think here we could leverage the `useAsyncState` to  determine whether to check via `AsyncKeyedStateBackend` or `AbstractKeyedStateBackend`.